torch/testing/_comparison.py

```diff
@@ -260,34 +260,40 @@ def make_tensor_mismatch_msg(
     """
 
     def unravel_flat_index(flat_index: int) -> Tuple[int, ...]:
-        if not mismatches.shape:
+        if not matches.shape:
             return ()
 
         inverse_index = []
-        for size in mismatches.shape[::-1]:
+        for size in matches.shape[::-1]:
             div, mod = divmod(flat_index, size)
             flat_index = div
             inverse_index.append(mod)
 
         return tuple(inverse_index[::-1])
 
-    number_of_elements = mismatches.numel()
-    total_mismatches = torch.sum(mismatches).item()
+    number_of_elements = matches.numel()
+    total_mismatches = number_of_elements - int(torch.sum(matches))
     extra = (
         f"Mismatched elements: {total_mismatches} / {number_of_elements} "
         f"({total_mismatches / number_of_elements:.1%})"
     )
 
-    a_flat = actual.flatten()
-    b_flat = expected.flatten()
-    matches_flat = ~mismatches.flatten()
+    actual_flat = actual.flatten()
+    expected_flat = expected.flatten()
+    matches_flat = matches.flatten()
 
-    abs_diff = torch.abs(a_flat - b_flat)
+    if not actual.dtype.is_floating_point and not actual.dtype.is_complex:
+        # TODO: Instead of always upcasting to int64, it would be sufficient to cast to the next higher dtype to avoid
+        #  overflow
+        actual_flat = actual_flat.to(torch.int64)
+        expected_flat = expected_flat.to(torch.int64)
+
+    abs_diff = torch.abs(actual_flat - expected_flat)
     # Ensure that only mismatches are used for the max_abs_diff computation
     abs_diff[matches_flat] = 0
     max_abs_diff, max_abs_diff_flat_idx = torch.max(abs_diff, 0)
 
-    rel_diff = abs_diff / torch.abs(b_flat)
+    rel_diff = abs_diff / torch.abs(expected_flat)
```

Comment:
A slight OT suggestion: could we have a better normalization factor here (say, `(torch.abs(actual) + torch.abs(expected)) / 2`) for the case where `expected` contains zeros (having zeros is typical, say when comparing the indices of sparse tensors)? Atm, the mismatch messages from `assert_close`  depends on the order of inputs, for example:
```
>>> torch.testing.assert_close(torch.tensor([1, 0]), torch.tensor([1, 1]))
<snip>
Mismatched elements: 1 / 2 (50.0%)
Greatest absolute difference: 1 at index (1,)
Greatest relative difference: 1.0 at index (1,)
>>> torch.testing.assert_close(torch.tensor([1, 1]), torch.tensor([1, 0]))
<snip>
Mismatched elements: 1 / 2 (50.0%)
Greatest absolute difference: 1 at index (1,)
Greatest relative difference: inf at index (1,)
```
(btw, reporting relative differences for non-float tensors is often pointless as well).

Reply:
> Atm, the mismatch messages from `assert_close` depends on the order of inputs, for example:

It's not just the messages, it is the actual op. Internally, we rely on `torch.isclose` and that is already asymmetric. It [defines closeness](https://pytorch.org/docs/stable/generated/torch.isclose.html?highlight=isclose#torch.isclose) as `abs(actual - expected) <= atol + rtol * abs(expected)`. Believe me when I say, we (`torch.testing` team) wanted to change that, but there is just too much inertia. PyTorch is not an outlier here; [`numpy` ](https://numpy.org/doc/stable/reference/generated/numpy.isclose.html#numpy-isclose) (and virtually every other array library) is doing the same.

[Pythons `math` module](https://docs.python.org/3/library/math.html?highlight=isclose#math.isclose) is doing the more sensible thing in defining closeness as `abs(actual - expected) <= max(atol, rtol * max(abs(actual), abs(expected)))`. You can read more about this whole issue in [PEP485](https://peps.python.org/pep-0485/).

At some point we tried to get this behavior specified by the Array API, but couldn't gain enough traction. See data-apis/array-api#170.

> (btw, reporting relative differences for non-float tensors is often pointless as well).

Doesn't that somewhat contradict the use case you gave earlier?

> for the case where expected contains zeros (having zeros is typical, say when comparing the indices of sparse tensors)